search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,568 @@
+    g_free (page_data);
+}
+
+static GVariant *
+get_result_metas (gchar     **page_ids,
+                  GHashTable *page_data)
+{
+    GVariantBuilder metas;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++) {
+        GVariantBuilder meta;
+        gchar *gicon_str;
+        PageData *data = g_hash_table_lookup (page_data, page_ids[i]);
+        g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+        g_variant_builder_add (&meta, "{sv}",
+                              "id", g_variant_new_string (page_ids[i]));
+        g_variant_builder_add (&meta, "{sv}",
+                               "name", g_variant_new_string (data->title));
+        gicon_str = g_icon_to_string (data->icon);
+        g_variant_builder_add (&meta, "{sv}",
+                              "gicon", g_variant_new_string (gicon_str));
+        g_free (gicon_str);
+        g_variant_builder_add (&meta, "{sv}",
+                               "description", g_variant_new_string (data->desc));
+        g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+
+static gboolean
+match_word (const gchar *search_phrase,
+            const gchar *title_casefold,
+            const gchar *desc_casefold)
+{
+    static GRegex *nonword = NULL;
+    gint stri;
+    gchar **strs;
+    gboolean ret = TRUE;
+
+    if (nonword == NULL)
+        nonword = g_regex_new ("\\W", 0, 0, NULL);
+    g_assert (nonword != NULL);
+
+    strs = g_regex_split (nonword, search_phrase, 0);
+    for (stri = 0; strs[stri]; stri++) {
+        if (!title_casefold || !strstr (title_casefold, strs[stri])) {
+            if (!desc_casefold || !strstr (desc_casefold, strs[stri])) {
+                ret = FALSE;
+                break;
+            }
+        }
+    }
+
+    g_strfreev (strs);
+
+    return ret;
+}
+
+static GVariant *
+get_search_results (gchar     **terms,
+                    GHashTable *page_data)
+{
+    GVariantBuilder results;
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    GPtrArray *res = g_ptr_array_new ();
+    gint i;
+    const gchar *page_id;
+    PageData *data;
+
+    g_hash_table_iter_init (&iter, page_data);
+    g_variant_builder_init (&results, G_VARIANT_TYPE ("as"));
+
+    while (g_hash_table_iter_next (&iter, (gpointer *) &page_id, (gpointer *) &data)) {
+        if (match_word (term, data->title_casefold, data->desc_casefold))
+            g_variant_builder_add (&results, "s", page_id);
+    }
+
+    g_free(term);
+
+    return g_variant_new ("(as)", &results);
+}
+
+typedef GVariant * (*ResultGetter) (gchar     **terms,
+                                    GHashTable *page_data);
+
+typedef struct
+{
+    YelpSearchProviderApp *app;
+    GDBusMethodInvocation *invocation;
+    gchar                **terms;
+    ResultGetter           result_getter;
+} DelayedInitialResultSet;
```

Comment:
Consider renaming it to `DelayedResultGetter` or something like this. It is not only about delayed getting of the initial results set.